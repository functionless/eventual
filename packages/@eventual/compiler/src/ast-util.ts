import {
  Argument,
  ArrowFunctionExpression,
  CallExpression,
  Expression,
  FunctionExpression,
  Span,
  Node,
  StringLiteral,
  FunctionDeclaration,
  Statement,
  HasSpan,
  Identifier,
  ComputedPropName,
} from "@swc/core";

/**
 * A heuristic for identifying a {@link CallExpression} that is a call to an API handler.
 *
 * 1. must be a call to a MemberExpression matching to `api.get` or
 *    `ev.api.get` where `get` is any of the allowed {@link apiCalls}.
 * 2. Must have between 2 and 3 arguments.
 */
export function isCommandCall(call: CallExpression): boolean {
  const c = call.callee;
  if (
    (c.type !== "Identifier" || c.value === "command") &&
    (c.type !== "MemberExpression" || isId(c.property, "command"))
  ) {
    return false;
  }

  return call.arguments.length === 2 || call.arguments.length === 3;
}

/**
 * A heuristic for identifying a {@link CallExpression} that is a call to an `onEvent` handler.
 *
 * 1. must be a call to a MemberExpression matching to `<expression>.onEvent(impl | props, impl)`.
 * 2. must have 1 or 2 arguments.
 */
export function isOnEventCall(call: CallExpression): boolean {
  const c = call.callee;
  if (c.type === "MemberExpression") {
    if (isId(c.property, "onEvent")) {
      // eventType.onEvent(async () => { })
      return call.arguments.length === 1 || call.arguments.length === 2;
    }
  }
  return false;
}

/**
 * Checks whether a {@link node} is an {@link Identifier} with a value of {@link Value}.
 */
function isId<Value extends string>(
  node: Expression | ComputedPropName,
  value: Value | Set<Value> | undefined
): node is Identifier & {
  value: Value;
} {
  return (
    node.type === "Identifier" &&
    (value === undefined ||
      (typeof value === "string"
        ? node.value === value
        : value.has(node.value as any)))
  );
}

/**
 * A heuristic for identifying a {@link CallExpression} that is a call
 * to the eventual.workflow utility:
 *
 * 1. must be a function call with exactly 2 arguments
 * 2. first argument is a string literal
 * 3. second argument is a FunctionExpression or ArrowFunctionExpression
 * 4. callee is an identifier `"workflow"` or `<identifier>.workflow`
 */
export function isWorkflowCall(call: CallExpression): call is CallExpression & {
  arguments: [
    Argument & { expression: StringLiteral },
    Argument & { expression: FunctionExpression | ArrowFunctionExpression }
  ];
} {
  return (
    isWorkflowCallee(call.callee) &&
    call.arguments[0]?.expression.type === "StringLiteral" &&
    ((call.arguments.length === 2 &&
      isNonGeneratorFunction(call.arguments[1]?.expression)) ||
      (call.arguments.length === 3 &&
        isNonGeneratorFunction(call.arguments[2]?.expression)))
  );
}

export function isNonGeneratorFunction(
  expr?: Expression
): expr is ArrowFunctionExpression | FunctionExpression {
  return (
    (expr?.type === "ArrowFunctionExpression" ||
      expr?.type === "FunctionExpression") &&
    !expr.generator
  );
}

export function isActivityCallee(callee: CallExpression["callee"]) {
  return isCallee("activity", callee);
}

export function isWorkflowCallee(callee: CallExpression["callee"]) {
  return isCallee("workflow", callee);
}

export function isCallee(
  type: "activity" | "workflow",
  callee: CallExpression["callee"]
) {
  return (
    (callee.type === "Identifier" && callee.value === type) ||
    (callee.type === "MemberExpression" &&
      callee.property.type === "Identifier" &&
      callee.property.value === type)
  );
}

export function isAsyncFunctionDecl(
  stmt: Statement
): stmt is FunctionDeclaration & { async: true } {
  return stmt.type === "FunctionDeclaration" && stmt.async;
}

export function hasSpan(expr: Node): expr is Node & HasSpan {
  return "span" in expr;
}

export function getSpan(expr: Node): Span {
  if (hasSpan(expr)) {
    return expr.span;
  } else {
    // this is only true for JSXExpressions which we should not encounter
    throw new Error(`cannot get span of ${expr.type}`);
  }
}
