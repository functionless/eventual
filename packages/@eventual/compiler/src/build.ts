import { constants } from "fs";
import fs from "fs/promises";
import path from "path";

export async function prepareOutDir(outDir: string, clean: boolean = true) {
  try {
    await fs.access(outDir, constants.F_OK);
    if (clean) {
      await cleanDir(outDir);
    }
  } catch {
    await fs.mkdir(outDir, {
      recursive: true,
    });
  }
}
export async function rmrf(file: string) {
  const stat = await fs.stat(file);
  if (stat.isDirectory()) {
    await cleanDir(file);
  } else {
    await fs.rm(file);
  }
}

export async function cleanDir(dir: string) {
  await Promise.all(
    (await fs.readdir(dir)).map((file) => rmrf(path.join(dir, file)))
  );
}
