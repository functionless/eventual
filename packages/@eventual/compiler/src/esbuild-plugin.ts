import esBuild from "esbuild";
import { Module, parseFile, print } from "@swc/core";
import path from "path";
import { OuterVisitor } from "./workflow-visitor.js";

export const eventualESPlugin: esBuild.Plugin = {
  name: "eventual",
  setup(build) {
    build.onLoad({ filter: /\.[mc]?[tj]s$/g }, async (args) => {
      // FYI: SWC erases comments: https://github.com/swc-project/swc/issues/6403
      const sourceModule = await parseFile(args.path, {
        syntax: "typescript",
      });

      const outerVisitor = new OuterVisitor();
      const transformedModule = outerVisitor.visitModule(sourceModule);

      // only format the module and return it if we found eventual functions to transform.
      if (outerVisitor.foundEventual) {
        const { code } = await printModule(transformedModule, args.path);

        return {
          contents: code,
          loader: "ts",
        };
      }
      return;
    });
  },
};

async function printModule(module: Module, filePath: string) {
  return await print(module, {
    //sourceFileName doesnt set up the sourcemap path the same way as transform does...
    sourceFileName: path.basename(filePath),
    //Instead these two are needed
    filename: path.basename(filePath),
    outputPath: path.dirname(filePath),
    //esbuild will extract these out later
    sourceMaps: "inline",
    jsc: {
      target: "es2022",
    },
  });
}
