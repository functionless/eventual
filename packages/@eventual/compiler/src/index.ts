export * from "./build-service.js";
export * from "./eventual-bundle.js";
export * from "./eventual-infer.js";
export * from "./build.js";
