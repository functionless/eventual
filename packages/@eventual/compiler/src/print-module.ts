import { Module, print } from "@swc/core";
import path from "path";

export async function printModule(module: Module, filePath: string) {
  return await print(module, {
    // sourceFileName doesnt set up the sourcemap path the same way as transform does...
    sourceFileName: path.basename(filePath),
    // Instead these two are needed
    filename: path.basename(filePath),
    outputPath: path.dirname(filePath),
    // esbuild will extract these out later
    sourceMaps: "inline",
    jsc: {
      target: "es2022",
    },
  });
}
