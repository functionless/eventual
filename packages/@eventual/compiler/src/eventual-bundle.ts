import fs from "fs/promises";
import path from "path";
import esbuild from "esbuild";
import { esbuildPluginAliasPath } from "esbuild-plugin-alias-path";
import { eventualESPlugin } from "./esbuild-plugin.js";
import { prepareOutDir } from "./build.js";
import { createRequire } from "module";

// @ts-ignore - ts is complaining about not having module:esnext even thought it is in tsconfig.json
const require = createRequire(import.meta.url);

/**
 * Bundle an eventual program
 * @param outDir Directory to bundle to
 * @param entry File containing the service
 * @returns Paths to orchestrator and activtyWorker output files
 */
export async function bundle(
  outDir: string,
  serviceEntry: string
): Promise<void> {
  console.log("Bundling:", outDir, serviceEntry);
  await prepareOutDir(outDir);

  await Promise.all([
    build({
      name: "orchestrator",
      outDir,
      injectedEntry: serviceEntry,
      entry: runtimeEntrypoint("orchestrator"),
      plugins: [eventualESPlugin],
    }),
    build({
      name: "activity",
      outDir,
      injectedEntry: serviceEntry,
      entry: runtimeEntrypoint("activity-worker"),
    }),
    build({
      name: "api",
      outDir,
      injectedEntry: serviceEntry,
      entry: runtimeEntrypoint("api-handler"),
    }),
    //This one is actually an api function
    build({
      name: "list-workflows",
      outDir,
      injectedEntry: serviceEntry,
      entry: runtimeEntrypoint("list-workflows"),
    }),
  ]);
}

export async function bundleService(outDir: string, entry: string) {
  await prepareOutDir(outDir);
  return build({
    outDir,
    entry,
    name: "service",
    plugins: [eventualESPlugin],
    //It's important that we use inline source maps for service, otherwise debugger fails to pick it up
    sourcemap: "inline",
  });
}

export function runtimeEntrypoint(name: string) {
  return path.join(
    require.resolve("@eventual/aws-runtime"),
    `../../esm/handlers/${name}.js`
  );
}

async function build({
  outDir,
  injectedEntry,
  name,
  entry,
  plugins,
  sourcemap,
}: {
  injectedEntry?: string;
  outDir: string;
  name: string;
  entry: string;
  plugins?: esbuild.Plugin[];
  sourcemap?: boolean | "inline";
}) {
  const outfile = path.join(outDir, `${name}/index.mjs`);
  const bundle = await esbuild.build({
    mainFields: ["module", "main"],
    sourcemap: sourcemap ?? true,
    plugins: [
      ...(injectedEntry
        ? [
            esbuildPluginAliasPath({
              alias: {
                ["@eventual/entry/injected"]: path.resolve(injectedEntry),
              },
            }),
          ]
        : []),
      ...(plugins ?? []),
    ],
    conditions: ["module", "import", "require"],
    // supported with NODE_18.x runtime
    // TODO: make this configurable.
    // external: ["@aws-sdk"],
    platform: "node",
    format: "esm",
    //Target for node 16
    target: "es2021",
    metafile: true,
    bundle: true,
    entryPoints: [path.resolve(entry)],
    banner: esmPolyfillRequireBanner(),
    outfile,
  });

  await writeEsBuildMetafile(
    bundle,
    path.resolve(outDir!, `${name}/meta.json`)
  );

  return outfile;
}

/**
 * Allows ESM module bundles to support dynamo requires when necessary.
 */
function esmPolyfillRequireBanner() {
  return {
    js: [
      `import { createRequire as topLevelCreateRequire } from 'module'`,
      `const require = topLevelCreateRequire(import.meta.url)`,
    ].join("\n"),
  };
}

function writeEsBuildMetafile(
  esbuildResult: esbuild.BuildResult & { metafile: esbuild.Metafile },
  path: string
) {
  return fs.writeFile(path, JSON.stringify(esbuildResult.metafile));
}
