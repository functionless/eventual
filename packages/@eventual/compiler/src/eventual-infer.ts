/**
 * This script imports a user's script and outputs a JSON object
 * to stdout containing all of the data that can be inferred.
 *
 * @see ServiceSpec
 */
import { generateSchema } from "@anatine/zod-openapi";
import {
  ServiceSpec,
  commands,
  entities,
  events,
  subscriptions,
  tasks,
  transactions,
  workflows,
  buckets,
  BucketStreamSpec,
} from "@eventual/core/internal";
import {
  CallExpression,
  ExportDeclaration,
  Expression,
  ModuleDeclaration,
  TsType,
  parseFile,
} from "@swc/core";
import { Visitor } from "@swc/core/Visitor.js";
import crypto from "crypto";
import esbuild from "esbuild";
import fs from "fs/promises";
import os from "os";
import path from "path";
import {
  getSpan,
  isCommandCall,
  isEntityStreamCall,
  isEntityStreamMemberCall,
  isOnEventCall,
  isSubscriptionCall,
  isTaskCall,
} from "./ast-util.js";
import { printModule } from "./print-module.js";

export async function infer(
  scriptName = process.argv[2],
  openApi: ServiceSpec["openApi"]
): Promise<ServiceSpec> {
  if (scriptName === undefined) {
    throw new Error(`scriptName undefined`);
  }

  const tmp = os.tmpdir();

  const bundle = await esbuild.build({
    mainFields: ["module", "main"],
    entryPoints: [scriptName],
    plugins: [inferPlugin],
    sourcemap: false,
    bundle: true,
    write: false,
    platform: "node",
  });

  const script = bundle.outputFiles[0]!.text;
  const hash = crypto.createHash("md5").update(script).digest("hex");
  scriptName = path.join(tmp, `${hash}.js`);
  await fs.writeFile(scriptName, script);

  await import(path.resolve(scriptName));

  const serviceSpec: ServiceSpec = inferFromMemory(openApi);

  console.log(JSON.stringify(serviceSpec));

  return serviceSpec;
}

export function inferFromMemory(openApi: ServiceSpec["openApi"]): ServiceSpec {
  return {
    workflows: [...workflows().keys()].map((n) => ({ name: n })),
    tasks: Object.values(tasks()).map((task) => ({
      name: task.name,
      sourceLocation: task.sourceLocation,
      options: task.options,
    })),
    events: Array.from(events().values()).map((event) => ({
      name: event.name,
      schema: event.schema ? generateSchema(event.schema) : undefined,
    })),
    subscriptions: subscriptions().map((e) => ({
      name: e.name,
      props: {
        memorySize: e.props?.memorySize,
        retryAttempts: e.props?.retryAttempts,
        handlerTimeout: e.props?.handlerTimeout,
      },
      sourceLocation: e.sourceLocation,
      filters: e.filters,
    })),
    commands: commands.map((command) => ({
      name: command.name,
      sourceLocation: command.sourceLocation,
      path: command.path,
      description: command.description,
      summary: command.summary,
      memorySize: command.memorySize,
      handlerTimeout: command.handlerTimeout,
      method: command.method,
      input: command.input ? generateSchema(command.input) : undefined,
      output: command.output ? generateSchema(command.output) : undefined,
      passThrough: command.passThrough,
      params: command.params,
      validate: command.validate,
      namespace: command.namespace,
    })),
    buckets: {
      buckets: [...buckets().values()].map((b) => ({
        name: b.name,
        streams: b.streams.map(
          (s) =>
            ({
              name: s.name,
              bucketName: s.bucketName,
              options: s.options,
              sourceLocation: s.sourceLocation,
            } satisfies BucketStreamSpec)
        ),
      })),
    },
    entities: {
      entities: [...entities().values()].map((d) => ({
        name: d.name,
        schema: d.schema ? generateSchema(d.schema) : undefined,
        streams: d.streams.map((s) => ({
          name: s.name,
          entityName: s.entityName,
          options: s.options,
          sourceLocation: s.sourceLocation,
        })),
      })),
    },
    transactions: [...transactions().values()].map((t) => ({
      name: t.name,
    })),
    openApi,
  };
}

export const inferPlugin: esbuild.Plugin = {
  name: "eventual",
  setup(build) {
    build.onLoad({ filter: /\.[mc]?[tj]s$/g }, async (args) => {
      // FYI: SWC erases comments: https://github.com/swc-project/swc/issues/6403
      const sourceModule = await parseFile(args.path, {
        syntax: "typescript",
      });

      const inferVisitor = new InferVisitor(args.path);
      const transformedModule = inferVisitor.visitModule(sourceModule);

      if (inferVisitor.didMutate) {
        const { code } = await printModule(transformedModule, args.path);

        return {
          contents: code,
          loader: "js",
        };
      }
      return undefined;
    });
  },
};

export class InferVisitor extends Visitor {
  public didMutate: boolean = false;

  private exportName: string | undefined;

  constructor(readonly fileName: string) {
    super();
  }

  public visitTsType(n: TsType): TsType {
    return n;
  }

  visitExportDeclaration(decl: ExportDeclaration): ModuleDeclaration {
    if (decl.declaration.type === "VariableDeclaration") {
      if (
        !decl.declaration.declare &&
        decl.declaration.declarations.length === 1
      ) {
        const varDecl = decl.declaration.declarations[0]!;
        if (varDecl.id.type === "Identifier") {
          if (varDecl.init?.type === "CallExpression") {
            this.exportName = varDecl.id.value;

            const call = this.visitCallExpression(varDecl.init);

            this.exportName = undefined;

            return {
              ...decl,
              declaration: {
                ...decl.declaration,
                declarations: [
                  {
                    ...varDecl,
                    init: call,
                  },
                ],
              },
            };
          }
        }
      }
    }
    return super.visitExportDeclaration(decl);
  }

  visitCallExpression(call: CallExpression): Expression {
    if (
      this.exportName &&
      (isCommandCall(call) ||
        isOnEventCall(call) ||
        isSubscriptionCall(call) ||
        isTaskCall(call) ||
        isEntityStreamMemberCall(call) ||
        isEntityStreamCall(call))
    ) {
      this.didMutate = true;

      return {
        ...call,
        arguments: [
          {
            expression: {
              type: "ObjectExpression",
              span: getSpan(call),
              properties: [
                {
                  type: "KeyValueProperty",
                  key: {
                    type: "Identifier",
                    optional: false,
                    span: getSpan(call),
                    value: "exportName",
                  },
                  value: {
                    type: "StringLiteral",
                    span: getSpan(call),
                    value: this.exportName,
                  },
                },
                {
                  type: "KeyValueProperty",
                  key: {
                    type: "Identifier",
                    optional: false,
                    span: getSpan(call),
                    value: "fileName",
                  },
                  value: {
                    type: "StringLiteral",
                    span: {
                      ctxt: 0,
                      end: 0,
                      start: 0,
                    },
                    value: this.fileName,
                  },
                },
              ],
            },
          },
          ...call.arguments,
        ],
      };
    }
    return super.visitCallExpression(call);
  }
}
