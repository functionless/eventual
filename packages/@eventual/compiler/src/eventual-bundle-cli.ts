import { bundle } from "./eventual-bundle.js";

export async function cli() {
  try {
    const [, , outDir, entry] = process.argv;
    if (!(outDir && entry)) {
      throw new Error(`Usage: eventual-build <out-dir> <entry-point>`);
    }
    await bundle(outDir, entry);
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}
