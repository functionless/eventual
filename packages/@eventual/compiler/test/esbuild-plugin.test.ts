import "jest";

import path from "path";
import esbuild from "esbuild";
import { eventualESPlugin } from "../src/esbuild-plugin";

describe("esbuild-plugin", () => {
  test("ts workflow", async () => {
    const bundle = await esbuild.build({
      mainFields: ["module", "main"],
      entryPoints: [path.resolve(__dirname, "..", "test-files", "workflow.ts")],
      sourcemap: false,
      plugins: [eventualESPlugin],
      bundle: true,
      write: false,
    });

    expect(
      bundle
        .outputFiles![0]?.text.split("\n")
        // HACK: filter out comment that is breaking the tests when run from VS Code
        // TODO: figure out why running vs code test is having trouble identifying the right
        //       tsconfig.test.json without a configuration at the root.
        // HINT: something to do with `.vscode/launch.json`
        .filter((line) => !line.includes("test-files/workflow.ts"))
        .join("\n")
    ).toMatchSnapshot();
  });

  test("ts not workflow", async () => {
    const bundle = await esbuild.build({
      mainFields: ["module", "main"],
      entryPoints: [
        path.resolve(__dirname, "..", "test-files", "not-workflow.ts"),
      ],
      sourcemap: false,
      plugins: [eventualESPlugin],
      bundle: true,
      write: false,
    });

    expect(bundle.outputFiles![0]?.text).toMatchSnapshot();
  });

  test("mts workflow", async () => {
    const bundle = await esbuild.build({
      mainFields: ["module", "main"],
      entryPoints: [
        path.resolve(__dirname, "..", "test-files", "workflow.mts"),
      ],
      sourcemap: false,
      plugins: [eventualESPlugin],
      bundle: true,
      write: false,
    });

    expect(bundle.outputFiles![0]?.text).toMatchSnapshot();
  });

  test("json file", async () => {
    const bundle = await esbuild.build({
      mainFields: ["module", "main"],
      entryPoints: [
        path.resolve(__dirname, "..", "test-files", "json-file.json"),
      ],
      sourcemap: false,
      plugins: [eventualESPlugin],
      bundle: true,
      write: false,
    });

    expect(bundle.outputFiles![0]?.text).toMatchSnapshot();
  });
});
