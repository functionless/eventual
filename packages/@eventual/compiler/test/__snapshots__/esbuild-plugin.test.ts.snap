// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`esbuild-plugin json file 1`] = `
"(() => {
  var json_file_default = {};
})();
"
`;

exports[`esbuild-plugin mts workflow 1`] = `
""use strict";
(() => {
  var doWork = activity("doWork", async (input) => input);
  var workflow_default = workflow("workflow", function* (input) {
    const items = yield doWork(input);
    yield $Eventual.all(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield $Eventual.all(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield $Eventual.allSettled(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield $Eventual.any(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield $Eventual.race(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
  });
})();
"
`;

exports[`esbuild-plugin open-account 1`] = `
""use strict";
(() => {
  var open_account_default = workflow("open-account", function* ({ accountId, address, email, bankDetails }) {
    const rollbacks = [];
    try {
      yield createAccount(accountId);
    } catch (err) {
      console.error(err);
      throw err;
    }
    try {
      yield addAddress(accountId, address);
      rollbacks.push($eventual(function* () {
        return removeAddress(accountId);
      }));
      yield addEmail(accountId, email);
      rollbacks.push($eventual(function* () {
        return removeEmail(accountId);
      }));
      yield addBankAccount(accountId, bankDetails);
      rollbacks.push($eventual(function* () {
        return removeBankAccount(accountId);
      }));
    } catch (err) {
      yield $Eventual.all(rollbacks.map((rollback) => rollback()));
    }
  });
})();
"
`;

exports[`esbuild-plugin ts not workflow 1`] = `
""use strict";
(() => {
  var doWork = async () => {
  };
  function not_workflow_default() {
    console.log(doWork);
  }
})();
"
`;

exports[`esbuild-plugin ts workflow 1`] = `
""use strict";
(() => {
  var doWork = activity("doWork", async (input) => input);
  var workflow_default = workflow("workflow", function* (input) {
    const items = yield doWork(input);
    yield $Eventual.all(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield $Eventual.all(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield $Eventual.allSettled(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield $Eventual.any(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield $Eventual.race(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    condition(() => true);
    const func = () => $Eventual.all(items.map($eventual(function* (item) {
      yield doWork(item);
    })));
    yield func();
    const func2 = $eventual(function* () {
      yield $Eventual.all(items.map($eventual(function* (item) {
        yield doWork(item);
      })));
    });
    yield func2();
  });
  var workflow2 = workflow("timeoutFlow", {
    timeoutSeconds: 100
  }, function* () {
    yield doWork("something");
  });
  var workflow3 = workflow("timeoutFlow", function* () {
    const callMe = $eventual(function* () {
      yield sleepFor(20);
    });
    yield callMe();
  });
})();
"
`;
